File: src/main/java/net/okocraft/suffix/Suffix.java
package net.okocraft.suffix;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandSender;
import org.bukkit.command.PluginCommand;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.util.StringUtil;

public class Suffix extends JavaPlugin {

	private FileConfiguration config;
	private FileConfiguration defaultConfig;

	@Override
	public void onEnable() {
		PluginCommand command = Objects.requireNonNull(getCommand("suffix"));
		command.setExecutor(this);
		command.setTabCompleter(this);
		saveDefaultConfig();

		config = getConfig();
		defaultConfig = getDefaultConfig();
	}

	@Override
	public void onDisable() {
	}

	@Override
	public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
		String suffix;
		Player target;

		if (sender.hasPermission("suffix.other")) {
			if (args.length == 0) {
				sender.sendMessage(getMessage("specify-player"));
				return false;
			}

			if (args.length == 1) {
				sender.sendMessage(getMessage("specify-suffix"));
				return false;
			}

			target = Bukkit.getPlayer(args[0]);
			if (target == null) {
				sender.sendMessage(getMessage("player-not-found"));
				return false;
			}

			suffix = args[1];

		} else {
			if (!(sender instanceof Player)) {
				sender.sendMessage(getMessage("player-only"));
				return false;
			}

			if (args.length == 0) {
				sender.sendMessage(getMessage("specify-suffix"));
				return false;
			}

			target = (Player) sender;
			suffix = args[0];
		}

		int maxLength = config.getInt("suffix-length", 1);
		int suffixLength = ChatColor.stripColor(ChatColor.translateAlternateColorCodes('&', suffix)).length();
		if (maxLength < suffixLength) {
			sender.sendMessage(getMessageReplaced("too-long-suffix", Map.of("%length%", String.valueOf(maxLength))));
			return false;
		}
		for (String blacklistPattern : config.getStringList("blacklist-pattern")) {
			if (suffix.contains(blacklistPattern)) {
				sender.sendMessage(getMessageReplaced("blacklist-pattern", Map.of("%pattern%", blacklistPattern)));
				return false;
			}
		}

		String sufixApplyCommand = config
				.getString("suffix-apply-command", "lp user %player_name% meta setsuffix 10 %suffix%")
				.replaceAll("%suffix%", suffix).replaceAll("%player_name%", target.getName());

		Bukkit.dispatchCommand(Bukkit.getConsoleSender(), sufixApplyCommand);

		return true;
	}

	@Override
	public List<String> onTabComplete(CommandSender sender, Command command, String alias, String[] args) {
		if (sender.hasPermission("suffix.other")) {
			if (args.length == 1) {
				return StringUtil.copyPartialMatches(args[0],
						Bukkit.getOnlinePlayers().stream().map(Player::getName).collect(Collectors.toList()),
						new ArrayList<>());
			}

			if (args.length == 2) {
				return StringUtil.copyPartialMatches(args[1], List.of("<suffix>"), new ArrayList<>());
			}
		} else {
			if (args.length == 1) {
				return StringUtil.copyPartialMatches(args[0], List.of("<suffix>"), new ArrayList<>());
			}
		}

		return List.of();
	}

	private FileConfiguration getDefaultConfig() {
		InputStream is = getResource("config.yml");
		return YamlConfiguration.loadConfiguration(new InputStreamReader(is));
	}

	private String getMessage(String key) {
		String fullKey = "messages." + key;
		return ChatColor.translateAlternateColorCodes('&', config.getString(fullKey, defaultConfig.getString(fullKey)));
	}

	private String getMessageReplaced(String key, Map<String, String> placeholders) {
		String result = getMessage(key);
		for (Map.Entry<String, String> placeholder : placeholders.entrySet()) {
			result = result.replaceAll(placeholder.getKey(), placeholder.getValue());
		}

		return result;
	}
}

File: src/main/kotlin/net/okocraft/suffix/Suffix.kt
package net.okocraft.suffix

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.command.TabExecutor
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import org.bukkit.util.StringUtil

class Suffix : JavaPlugin(), CommandExecutor, TabExecutor {

    override fun onEnable() {
        saveDefaultConfig()
        getCommand("suffix")?.let {
            it.setExecutor(this)
            it.tabCompleter = this
        }
    }

    /**
     * /suffix player suffix
     * /suffix suffix
     */
    override fun onCommand(sender: CommandSender, command: Command, label: String, args: Array<out String>): Boolean {

        if (args.isEmpty()) {
            return false
        }

        if (!sender.hasPermission("suffix.change")) {
            sender.sendMessage(config.getString("messages.no-permission", "")!!)
            return false
        }

        val player: Player
        var suffix: String

        when {
            args.size > 1 -> {
                if (!sender.hasPermission("suffix.change.other")) {
                    sender.sendMessage(config.getString("messages.no-permission", "")!!)
                }

                if (Bukkit.getPlayer(args[0]) == null) {
                    sender.sendMessage(config.getString("messages.player-is-offline", "")!!)
                    return false
                }

                player = Bukkit.getPlayer(args[0])!!
                suffix = ChatColor.translateAlternateColorCodes('&', args[1])
            }
            sender is Player -> {
                player = sender
                suffix = ChatColor.translateAlternateColorCodes('&', args[0])
            }
            else -> {
                sender.sendMessage(config.getString("messages.specify-player", "")!!)
                return false
            }
        }

        val suffixChar = ChatColor.stripColor(suffix)!![0]
        suffix = suffix.substringBefore(suffixChar) + suffixChar
        val suffixCommand = config.getString("suffix-command", "")!!
            .replace("%player%", player.name)
            .replace("%suffix%", suffix)

        return Bukkit.dispatchCommand(Bukkit.getConsoleSender(), suffixCommand)
    }

    override fun onTabComplete(sender: CommandSender, command: Command, alias: String, args: Array<out String>): MutableList<String> {
        if (!sender.hasPermission("suffix.change.other")) {
            val players = Bukkit.getOnlinePlayers().toList().map { it.name }.toMutableList()
            if (args.size == 1) {
                return StringUtil.copyPartialMatches(args[0], players, mutableListOf())
            }

            if (args.size == 2) {
                return mutableListOf("<suffix>")
            }

            return mutableListOf()
        }

        if (args.size == 1) {
            return mutableListOf("<suffix>")
        }

        return mutableListOf()
    }
}